src/libre/re_analysis.c

```diff
@@ -5,78 +5,645 @@
+		}
+		return res;
+	}
+
+	/* Next passes, mark all nodes in a
+	 * first and/or last position. */
+        {
+		assign_firsts(ast->expr);
+		assign_lasts(ast->expr);
+	}
+
+	/* Next pass: set anchoring, now that nullability info from
+	 * the first pass is in place and some other things have been
+	 * cleaned up, and note whether the regex has any unsatisfiable
+	 * start anchors.*/
+        {
+		struct anchoring_env env;
+		memset(&env, 0x00, sizeof(env));
+		res = analysis_iter_anchoring(&env, ast->expr);
+		if (res != RE_ANALYSIS_OK) { return res; }
+	}
+
+	return res;
 }
 
-/* Now that the AST is complete, do a pass filling in details. */
-/* TODO: extract traversal from the callback, so the individual
- * analysis passes have an obvious focus. */
-static void
+static unsigned
+count_chain(const struct ast_expr *n, enum ast_expr_type t)
+{
+	unsigned res = 0;
+	for (;;) {
+		assert(n != NULL);
+		if (n->t == AST_EXPR_EMPTY) {
+			return res;
+		} else {
+			assert(n->t == t);
+			res++;
+			switch (t) {
+			case AST_EXPR_CONCAT:
+				n = n->u.concat.r;
+				break;
+			case AST_EXPR_ALT:
+				n = n->u.alt.r;
+				break;
+			default:
+				assert(0);
+				break;
+			}
+		}
+	}
+}
+
+static struct ast_expr *
+collect_chain(size_t count, struct ast_expr *doomed)
+{
+	size_t i;
+	if (doomed->t == AST_EXPR_CONCAT) {
+		if (count == 1) {
+			struct ast_expr *res = doomed->u.concat.l;
+			if (!flatten(&doomed->u.concat.l)) { return 0; }
+			res = doomed->u.concat.l;
+
+			assert(doomed->u.concat.r->t == AST_EXPR_EMPTY);
+			doomed->u.concat.l = re_ast_expr_tombstone();
+			re_ast_expr_free(doomed);
+			assert(res->t != AST_EXPR_CONCAT);
+			return res;
+		} else {
+			struct ast_expr *dst = re_ast_expr_concat_n(count);
+			if (dst == NULL) { return NULL; }
+			
+			for (i = 0; i < count; i++) {
+				struct ast_expr *ndoomed = doomed->u.concat.r;
+				if (!flatten(&doomed->u.concat.l)) { return 0; }
+				dst->u.concat_n.n[i] = doomed->u.concat.l;
+				doomed->u.concat.l = re_ast_expr_tombstone();
+				doomed->u.concat.r = re_ast_expr_tombstone();
+				re_ast_expr_free(doomed);
+				doomed = ndoomed;
+				if (i == count - 1) {
+					assert(doomed->t == AST_EXPR_EMPTY);
+				}
+			}
+			return dst;
+		}
+	} else if (doomed->t == AST_EXPR_ALT) {
+		if (count == 1) {
+			/* If we get here, it's a parser bug. Right? */
+			assert(0);
+		} else {
+			struct ast_expr *dst = re_ast_expr_alt_n(count);
+			if (dst == NULL) { return NULL; }
+
+			for (i = 0; i < count; i++) {
+				struct ast_expr *ndoomed = doomed->u.alt.r;
+				if (!flatten(&doomed->u.alt.l)) { return 0; }
+				dst->u.alt_n.n[i] = doomed->u.alt.l;
+				doomed->u.alt.l = re_ast_expr_tombstone();
+				doomed->u.alt.r = re_ast_expr_tombstone();
+				re_ast_expr_free(doomed);
+				doomed = ndoomed;
+				if (i == count - 1) {
+					assert(doomed->t == AST_EXPR_EMPTY);
+				}
+			}
+			return dst;
+		}
+	} else {
+		assert(0);
+		return NULL;
+	}
+}
+
+/* TODO: do this directly in the parser. */
```

Comment:
Would that get rid of the linked-list AST node type?

Reply:
Yes, and the linked-list ALT type as well.